test/distributed/optim/test_named_optimizer.py

```diff
@@ -0,0 +1,189 @@
+# Owner(s): ["oncall: distributed"]
+
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import unittest
+
+import torch
+import torch.nn as nn
+
+from torch.distributed.optim import NamedOptimizer
+
+
+class TestDummyModel(torch.nn.Module):
+    def __init__(self):
+        super(TestDummyModel, self).__init__()
+        torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+
+class NamedOptimizerTest(unittest.TestCase):
+    def _compare_state_dict_group(self, group, named_group, compare_equal=True):
+        for key, val in group.items():
+            if key != "params":
+                assert key in named_group, f"{key} not in named optimizer state dict"
+                if isinstance(val, torch.Tensor):
+                    if compare_equal:
+                        assert torch.allclose(val, named_group[key]), f"{key} state not equal"
+                    else:
+                        assert not torch.allclose(val, named_group[key]), f"{key} state equal"
+                else:
+                    if compare_equal:
+                        assert val == named_group[key], f"{key} state not equal"
+                    else:
+                        assert val != named_group[key], f"{key} state equal"
+
+
+    def test_state_dict(self):
+        """Check that NamedOptimizer exposes the expected state dict
+        interface."""
+        m = TestDummyModel()
+        m_dup = TestDummyModel()
+        optim_1 = torch.optim.SGD(
+            [
+                {"params": m.net1.parameters()},
+                {"params": m.net3.parameters(), "lr": 1e-3},
+            ],
+            lr=1e-2,
+            momentum=0.9,
+        )
+
+        optim_2 = torch.optim.Adam(
+            [
+                {"params": m.net2.parameters()},
+                {"params": m.net4.parameters(), "lr": 1e-5},
+            ]
+        )
+
+        named_optim_1 = NamedOptimizer(
+            m_dup.named_parameters(),
+            torch.optim.SGD,
+            [
+                {"params": m_dup.net1.parameters()},
+                {"params": m_dup.net3.parameters(), "lr": 1e-3},
+            ],
+            lr=1e-2,
+            momentum=0.9,
+        )
+
+        named_optim_2 = NamedOptimizer(
+            m_dup.named_parameters(),
+            torch.optim.Adam,
+            [
+                {"params": m_dup.net2.parameters()},
+                {"params": m_dup.net4.parameters(), "lr": 1e-5},
+            ],
+        )
+        for i in range(10):
```

Comment:
nit: Do we really need more than 1 iteration to create optimizer state?

Maybe change this to just `2` like you do in `test_load_state_dict` to save some computation? (not that the computation is the bottleneck though)

Reply:
Oh right.. I previously thought it might change after couple iterations. But I agree that 2 is enough here.